### PR TITLE
Feature patterns parsing

### PR DESCRIPTION
**Pull request by lyzadanger**
Hi! This PR introduces some features around `rendering` `patterns`, that is, constructing an object that represents what should ultimately be output as HTML pages. It does not handle the actual writing to the file system yet; to keep this PR from getting too huge I'd like to do that separately.

The main point: a directory under the main `patterns` directory that contains direct-descendent pattern files is considered a `collection` and will generate an output HTML page using a template that can iterate over the collection's patterns and output them. This PR handles the rendering part of that—compiling the right templates with the right context—but not the filesystem output yet.

When reviewing this, it may be helpful to know/remember that `drizzle` executes in four _phases_:

* `prepare` — Primarily concerned with templating prep—parsing and registering partials and helpers and whatnot
* `parse` — Parse and process different kinds of source files (patterns, data, pages...)
* `render` — Compile templates and build objects that represent output
* `write` — Write rendered resources to fs

This PR is primarily concerned with the `render` phase for `patterns` but extends some other areas as needed to support that.

Here are some things in this PR:
* Adds layout parsing to the `prepare` step — this registers layouts as partials. Necessary for `handlebars-layouts` and other reasons.
* Adjusts the `parse` step for `patterns`
* Adjusts the `parseLocalData` function (tweaks precedence of different data sources for a given resource)
* Adds a `render` module for `patterns`. Solidifies the notion of `pattern` `collections`
* Adds a `prepare` step for `patterns` to register each as a `partial`
* Adds and adjusts some functions in `utils` related to object organization and resource IDs
* Adds the notion of a `patternCollection` template (name may change to be simpler) used to render collections of patterns. Also adds fixtures for same.
* Adds additional fixtures and a bunch of tests.

/cc @tylersticka @erikjung

### CODE REVIEW

File: src/render/patterns.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * Patterns will not render individual pages. This function mutates `patterns`.
+ *
+ * @param {Object} patterns      The current level of the patterns tree we're
+ *                               rendering
+ * @param {Object} drizzleData   All the data we have, including `options`
+ * @param {String} collectionKey The key of the current set of `patterns`.
+ *                               Used to derive the collection's "name"
+ */
+function renderPatternCollection (patterns, drizzleData, collectionKey) {
+  for (var individualPatternKey in patterns.items) {
```
Comment (mrgerardorodriguez) [inline src/render/patterns.js]: Might this be a good place to use `let` over `var`?

File: src/render/patterns.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    // We want to render only pattern-collection pages, not individual
+    // patterns. Remove `contents` from individual patterns so the write
+    // phase doesn't create files for them.
+    delete patterns.items[individualPatternKey].contents;
+  }
+  patterns.name = utils.titleCase(utils.keyname(collectionKey));
+  patterns.contents = renderUtils.applyTemplate(
+    drizzleData.layouts.patternCollection.contents, // TODO obviously fragile
```
Comment (mrgerardorodriguez) [inline src/render/patterns.js]: Do you mind expanding on your thought of why this is fragile? Curious to learn. :)

File: src/render/patterns.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * for "pattern collections" (any entry in `patterns` that contains an
+ * `items` property).
+ *
+ * @param {Object} patterns     The current level of the `patterns` object.
+ * @param {Object} drizzleData
+ * @param {String} currentKey   The key for the current `patterns` object.
+ * @return {Object} drizzleData All drizzleData
+ */
+function walkPatterns (patterns, drizzleData, currentKey = 'patterns') {
+  for (var patternKey in patterns) {
```
Comment (mrgerardorodriguez) [inline src/render/patterns.js]: Perhaps also another `let` candidate?

**Comment by mrgerardorodriguez** [comment]
Super tiny comments, though, nothing that should stop you from moving forward. :) LGTM! :+1:

File: src/render/patterns.js (hunk 2 above)
Comment (lyzadanger) [inline src/render/patterns.js]: I'm referencing a deep-nested object without testing to make sure that it's defined. It's temporary until I clean up how I want to handle identifying the right template for pattern collections.

**Comment by lyzadanger** [comment]
@mrgerardorodriguez As those variables are never altered after assignment, I opted for `const` (not least because it will make my linter not yell at me :) ).

**Pull request merged**